internal/directives/http_requester.go

```diff
@@ -0,0 +1,253 @@
+	}
+	return h.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+// validate validates httpRequester configuration against a JSON schema.
+func (h *httpRequester) validate(cfg Config) error {
+	return validate(h.schemaLoader, gojsonschema.NewGoLoader(cfg), h.Name())
+}
+
+func (h *httpRequester) runPromotionStep(
+	_ context.Context,
+	_ *PromotionStepContext,
+	cfg HTTPConfig,
+) (PromotionStepResult, error) {
+	req, err := h.buildRequest(cfg)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error building HTTP request: %w", err)
+	}
+	resp, err := h.getClient(cfg).Do(req)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error sending HTTP request: %w", err)
+	}
+	defer resp.Body.Close()
+	env, err := h.buildExprEnv(resp)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error building expression context from HTTP response: %w", err)
+	}
+	success, err := h.wasRequestSuccessful(cfg, resp.StatusCode, env)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error evaluating success criteria: %w", err)
+	}
+	failure, err := h.didRequestFail(cfg, resp.StatusCode, env)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error evaluating failure criteria: %w", err)
+	}
+	switch {
+	case success && !failure:
+		outputs, err := h.buildOutputs(cfg.Outputs, env)
+		if err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("error extracting outputs from HTTP response: %w", err)
+		}
+		return PromotionStepResult{
+			Status: kargoapi.PromotionPhaseSucceeded,
+			Output: outputs,
+		}, nil
+	case failure:
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseFailed},
+			fmt.Errorf("HTTP response met failure criteria")
+	default:
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseRunning}, nil
+	}
+}
+
+func (h *httpRequester) buildRequest(cfg HTTPConfig) (*http.Request, error) {
+	method := cfg.Method
+	if method == "" {
+		method = http.MethodGet
+	}
+	req, err := http.NewRequest(method, cfg.URL, bytes.NewBufferString(cfg.Body))
+	if err != nil {
+		return nil, fmt.Errorf("error creating HTTP request: %w", err)
+	}
+	for _, header := range cfg.Headers {
+		req.Header.Add(header.Name, header.Value)
+	}
+	if len(cfg.QueryParams) > 0 {
+		q := req.URL.Query()
+		for _, queryParam := range cfg.QueryParams {
+			q.Add(queryParam.Name, queryParam.Value)
+		}
+		req.URL.RawQuery = q.Encode()
+	}
+	return req, nil
+}
+
+func (h *httpRequester) getClient(cfg HTTPConfig) *http.Client {
+	httpTransport := cleanhttp.DefaultTransport()
+	if cfg.InsecureSkipTLSVerify {
+		httpTransport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true, // nolint: gosec
+		}
+	}
+	return &http.Client{
+		Transport: httpTransport,
```

Comment:
Should we allow a timeout to be configured?

Reply:
Tentatively fixed in a732823e5a2dea322822b73898e6e63208b55f2f

lmk what you think of the default and the limits.

Reply:
I would personally make it a `time.Duration` in the API, instead of a "timeout in seconds". But no really strong opinion on it.

Reply:
That was my preference as well, but...

There's no native JS/JSON duration type, so we'd end up having to accept a string representation of the duration. Apart from having to document the format, this makes validation considerably harder. You end up stuck having to do it with a relatively complex regex that is hard to update it/when we ever raise the max timeout.

Overcoming all of the above didn't seem worthwhile when when an integer was so easy to validate.

All that being said, if you know an easy/conventional way of doing this that I am unaware of, I would be quite happy for this to be a duration instead.

Reply:
I think those observations are fair, and I have no alternative for it right away.